File: samples/grpc/iot-service-scala/src/main/scala/iot/temperature/TemperatureEvents.scala
package iot.temperature

import scala.concurrent.Future

import akka.actor.typed.ActorSystem
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.cluster.sharding.typed.scaladsl.ShardedDaemonProcess
import akka.http.scaladsl.model.HttpRequest
import akka.http.scaladsl.model.HttpResponse
import akka.persistence.query.Offset
import akka.persistence.query.typed.EventEnvelope
import akka.persistence.r2dbc.query.scaladsl.R2dbcReadJournal
import akka.persistence.typed.PersistenceId
import akka.projection.Projection
import akka.projection.ProjectionBehavior
import akka.projection.ProjectionId
import akka.projection.eventsourced.scaladsl.EventSourcedProvider
import akka.projection.grpc.consumer.scaladsl.EventProducerPushDestination
import akka.projection.r2dbc.scaladsl.R2dbcProjection
import akka.projection.scaladsl.Handler
import akka.projection.scaladsl.SourceProvider
import akka.util.Timeout
import org.slf4j.LoggerFactory

/**
 * Handle temperature events pushed by the edge systems.
 */
object TemperatureEvents {
  val logger = LoggerFactory.getLogger("iot.temperature.TemperatureEvents")

  val TemperatureEventsStreamId = "temperature-events"

  private val ProducerEntityType = "Sensor"

  def pushedEventsGrpcHandler(implicit system: ActorSystem[_])
      : PartialFunction[HttpRequest, Future[HttpResponse]] = {

    val destination = EventProducerPushDestination(
      TemperatureEventsStreamId,
      proto.TemperatureEventsProto.javaDescriptor.getFile :: Nil)

    EventProducerPushDestination
      .grpcServiceHandler(destination)(system)
  }

  def initPushedEventsConsumer(implicit system: ActorSystem[_]): Unit = {

    implicit val askTimeout: Timeout = Timeout.create(
      system.settings.config.getDuration("iot-service.ask-timeout"))

    val sharding = ClusterSharding(system)

    def sourceProvider(sliceRange: Range)
        : SourceProvider[Offset, EventEnvelope[proto.TemperatureRead]] =
      EventSourcedProvider.eventsBySlices[proto.TemperatureRead](
        system,
        readJournalPluginId = R2dbcReadJournal.Identifier,
        ProducerEntityType,
        sliceRange.min,
        sliceRange.max)

    def projection(
        sliceRange: Range): Projection[EventEnvelope[proto.TemperatureRead]] = {
      val minSlice = sliceRange.min
      val maxSlice = sliceRange.max
      val projectionId =
        ProjectionId("TemperatureEvents", s"temperature-$minSlice-$maxSlice")

      val handler: Handler[EventEnvelope[proto.TemperatureRead]] = {
        (envelope: EventEnvelope[proto.TemperatureRead]) =>
          logger.info(
            "Saw projected event: {} #{}: {}",
            envelope.persistenceId,
            envelope.sequenceNr,
            envelope.eventOption)

          envelope.event match {
            case proto.TemperatureRead(temperature, _) =>
              val entityId =
                PersistenceId.extractEntityId(envelope.persistenceId)
              val entityRef =
                sharding.entityRefFor(SensorTwin.EntityKey, entityId)
              entityRef.askWithStatus(
                SensorTwin.UpdateTemperature(temperature, _))
            case unknown =>
              throw new RuntimeException(
                s"Unknown event type: ${unknown.getClass}")
          }
      }

      R2dbcProjection.atLeastOnceAsync(
        projectionId,
        settings = None,
        sourceProvider(sliceRange),
        handler = () => handler)
    }

    // Split the slices into N ranges
    val numberOfSliceRanges: Int = system.settings.config
      .getInt("iot-service.temperature.projections-slice-count")
    val sliceRanges = EventSourcedProvider.sliceRanges(
      system,
      R2dbcReadJournal.Identifier,
      numberOfSliceRanges)

    ShardedDaemonProcess(system).init(
      name = "TemperatureProjection",
      numberOfInstances = sliceRanges.size,
      behaviorFactory = i => ProjectionBehavior(projection(sliceRanges(i))),
      stopMessage = ProjectionBehavior.Stop)

  }

}

File: samples/grpc/iot-service-java/src/main/java/iot/temperature/TemperatureEvents.java
package iot.temperature;

import akka.Done;
import akka.actor.typed.ActorSystem;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.ShardedDaemonProcess;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.japi.Pair;
import akka.japi.function.Function;
import akka.persistence.query.Offset;
import akka.persistence.query.typed.EventEnvelope;
import akka.persistence.r2dbc.query.javadsl.R2dbcReadJournal;
import akka.persistence.typed.PersistenceId;
import akka.projection.Projection;
import akka.projection.ProjectionBehavior;
import akka.projection.ProjectionId;
import akka.projection.eventsourced.javadsl.EventSourcedProvider;
import akka.projection.grpc.consumer.javadsl.EventProducerPushDestination;
import akka.projection.grpc.consumer.javadsl.Transformation;
import akka.projection.javadsl.Handler;
import akka.projection.javadsl.SourceProvider;
import akka.projection.r2dbc.javadsl.R2dbcProjection;
import iot.temperature.proto.TemperatureRead;
import java.time.Duration;
import java.util.Collections;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class TemperatureEvents {

  private static final Logger logger = LoggerFactory.getLogger(TemperatureEvents.class);

  public static final String TEMPERATURE_EVENT_STREAM_ID = "temperature-events";

  private static final String PRODUCER_ENTITY_TYPE = "Sensor";

  public static Function<HttpRequest, CompletionStage<HttpResponse>> pushedEventsGrpcHandler(
      ActorSystem<?> system) {
    var destination =
        EventProducerPushDestination.create(
                TEMPERATURE_EVENT_STREAM_ID,
                Collections.singletonList(iot.temperature.proto.TemperatureEvents.getDescriptor()),
                system);

    return EventProducerPushDestination.grpcServiceHandler(destination, system);
  }

  private static class TemperatureReadHandler extends Handler<EventEnvelope<TemperatureRead>> {

    private final ClusterSharding sharding;
    private final Duration askTimeout;

    public TemperatureReadHandler(ActorSystem<?> system) {
      this.sharding = ClusterSharding.get(system);
      this.askTimeout = system.settings().config().getDuration("iot-service.ask-timeout");
    }

    @Override
    public CompletionStage<Done> process(EventEnvelope<TemperatureRead> envelope) {
      logger.info(
          "Saw projected event: {}-{}: {}",
          envelope.persistenceId(),
          envelope.sequenceNr(),
          envelope.eventOption());

      var entityId = PersistenceId.extractEntityId(envelope.persistenceId());

      var entityRef = sharding.entityRefFor(SensorTwin.ENTITY_KEY, entityId);

      return entityRef.askWithStatus(
          replyTo -> new SensorTwin.UpdateTemperature(envelope.event().getTemperature(), replyTo),
          askTimeout);
    }
  }

  public static void initPushedEventsConsumer(ActorSystem<?> system) {
    // Split the slices into N ranges
    var numberOfSliceRanges =
        system.settings().config().getInt("iot-service.temperature.projections-slice-count");

    var sliceRanges =
        EventSourcedProvider.sliceRanges(
            system, R2dbcReadJournal.Identifier(), numberOfSliceRanges);

    ShardedDaemonProcess.get(system)
        .init(
            ProjectionBehavior.Command.class,
            "TemperatureProjection",
            sliceRanges.size(),
            i -> ProjectionBehavior.create(projection(system, sliceRanges.get(i))),
            ProjectionBehavior.stopMessage());
  }

  private static Projection<EventEnvelope<TemperatureRead>> projection(
      ActorSystem<?> system, Pair<Integer, Integer> sliceRange) {
    var minSlice = sliceRange.first();
    var maxSlice = sliceRange.second();
    var projectionId =
        ProjectionId.of("TemperatureEvents", "temperature-" + minSlice + "-" + maxSlice);

    SourceProvider<Offset, EventEnvelope<TemperatureRead>> sourceProvider =
        EventSourcedProvider.eventsBySlices(
            system,
            R2dbcReadJournal.Identifier(),
            PRODUCER_ENTITY_TYPE,
            sliceRange.first(),
            sliceRange.second());

    return R2dbcProjection.atLeastOnceAsync(
        projectionId,
        Optional.empty(),
        sourceProvider,
        () -> new TemperatureReadHandler(system),
        system);
  }
}
